exploration/0003-formatted-parts.md

```diff
@@ -0,0 +1,187 @@
+		<dd>@eemeli</dd>
+		<dt>First proposed</dt>
+		<dd>2023-08-29</dd>
+		<dt>Pull Request</dt>
+		<dd>#000</dd>
+	</dl>
+</details>
+
+## Objective
+
+Messages often include structures and values
+that are not necessarily best represented by a plain concatenated string.
+
+It would be useful to define a formatted-parts target for MessageFormat 2.
+
+## Background
+
+Past examples have shown us that if we don't provide a formatter to parts,
+the string output will be re-parsed and re-processed by users.
+
+## Use-Cases
+
+- Markup elements
+- Non-string values
+- Message post-processors
+
+## Requirements
+
+- Define an iterable sequence of formatted part objects.
+- Include metadata for each part, such as type, source, direction, and locale.
+- Allow the representation of non-string values.
+- Allow the representation of values that consist of an iterable sequence of formatted parts.
+- Be able to represent each resolved value of a pattern with any number of formatted parts, including none.
+- Define the formatted parts in a manner that allows synonymous but appropriate implementations in different programming languages.
+
+## Constraints
+
+- The JS Intl formatters already include formatted-parts representations for each supported data type.
+  The JS implementation of the MF2 formatted-parts representation should be able to match their structure,
+  at least as far as that's possible and appropriate.
+
+## Proposed Design
+
+The formatted-parts API is included in the spec as an optional but recommended formatting target.
+
+The shape of the formatted-parts output is defined in a manner similar to the data model,
+which includes TypeScript, JSON Schema, and XML DTD definitions of the same data structure.
+
+At the top level, the formatted-parts result is an iterable sequence of parts.
+Parts corresponding to each _text_ can be simpler than those of _expressions_,
+as they do not have a `source` other than their `value`,
+or set any of the other possible metadata fields.
+
+```ts
+type MessageParts = Iterable<
+  MessageTextPart | MessageExpressionPart | MessageBiDiIsolationPart
+>;
+
+interface MessageTextPart {
+  type: "text";
+  value: string;
+}
+```
+
+For MessageExpressionPart, the `source` corresponds to the expression's fallback value.
+The parts' `dir` and `locale` may be defined if overridden by an expression attribute or set otherwise.
+Each part should have at most one of `value` or `parts` defined;
+some may have none.
+
+```ts
+interface MessageExpressionPart {
+  type: string;
+  source: string;
+  parts?: Iterable<{ type: string; value: unknown }>;
+  value?: unknown;
+  dir?: "ltr" | "rtl" | "auto";
+  locale?: string;
+}
+```
+
+The bidi isolation strategies included in the spec may require
+the insertion of MessageBiDiIsolationParts in the formatted-parts output.
+
+```ts
+interface MessageBiDiIsolationPart {
+  type: "bidiIsolation";
+  value: "\u2066" | "\u2067" | "\u2068" | "\u2069"; // LRI | RLI | FSI | PDI
+}
+```
```

Comment:
Maybe? But since the parts have a direction, the bidi isolation parts are not necessarily required. Let's consider pattern:

```
{Today is {$date :datetime}}
```

In locale `ar-AE`, the output of the `:datetime` function might be "30‏/08‏/2023" with a `dir` of `rtl`. Note that the string looks bad here because it is an LTR context and not bidi isolated. There are U+200F RLMs after `08` and `30` to assist, but the string needs to be wrapped.

Ignoring that the date has interior parts, getting a MessageExpressionPart:

```
value: "30‏/08‏/2023"
dir: "rtl"
```

... could result in a formatted string (in an HTML context) like:

```
Today is <span dir=rtl>30‏/08‏/2023</span>.
```

This draws correctly without any further intervention.

If this became a list of parts it still works:

```
parts: [
   { name: "day", value: "30\u200f" },
   { name: "sep", value: "/"},
   { name: "month", value: "08\u200f"},
   { name: "sep", value: "/"},
   { name: "year", value: "2023"}
]
dir: "rtl"
```

The caller could use this to produce:

```html
Today is <span dir=rtl>
<span class="day">30&#x200f;</span>/08&#x200f;2023</span>.
```

[Test](https://w3c.github.io/i18n-discuss/explainers/bidi-html-demo.html?text=Today%20is%20%3Cspan%20class=magenta%20dir=rtl%3E30\u200f/08\u200f/2023%3C/span%3E.)

You wouldn't want Unicode isolates in this case.

Reply:
I agree that in some cases bidi isolation of formatted-parts output may be achieved e.g. via `<span dir=...>` tags, but I don't think that's universally true for every use case for formatted parts.

This is one reason why e.g. in the JS Intl.MF proposal my proposal for bidi handling includes the `"none"` strategy as an alternative to the default (and required) `"compatibility"` strategy.

Effectively, I think we should include the MessageBiDiIsolationPart by default, but allow for an implementation to provide a way to get rid of them, or just have a consumer of the output ignore all the `type: "bidiIsolation"` parts.

Reply:
I agree that span is not the only way that bidi will be implemented. `none` is easy: one can just ignore the bidi/direction. Obviously sometimes folks will want bidi controls. Or they might want to make an attributed string, I suppose, in some UI frameworks. 

I'm reluctant to create a "bidi isolation part" because it separates the direction information from the value it is associated with. It's harder if one is doing decoration, such as building some control using various tags, not all of which are span, some of which have classes or other attributes. The only place where the isolation part is easier is stringifying the message (since one just consumes it. 

```javascript
function stringifyParts(parts) {
  let res = ''
  for (let part of parts) {
    res += bidiStrategyImpl.start(part.dir);  // none does nothing, compatibility does a control, etc.
    if (part.type === 'fallback') res += `{${part.source}}`
    else if ('value' in part) res += String(part.value)
    else if ('parts' in part) {
      for (let sub of part.parts) res += String(sub.value)
    }
    res += bidiStrategyImpl.end(part.dir);
  }
  return res
}
```

Reply:
Doing as you propose would require implementations to provide explicit `bidiStrategyImpl.start()` and `bidiStrategyImpl.end()` interfaces, or the behaviour of format-to-parts-to-string could not be guaranteed to match the behaviour of format-to-string.

That seems like a much bigger ask than defining what the bidi isolation parts would look like if they are to be included in the parts output by the requested bidi isolation strategy.

Reply:
Format-to-string's output would depend on what the bidi isolation strategy was, no? 

The start/end thing was just an example. Note that with isolation parts you have to "read ahead" to find out what is being isolated if what one is doing with the parts is decorating a control or generating HTML.

Reply:
> Format-to-string's output would depend on what the bidi isolation strategy was, no?

Yes.

> The start/end thing was just an example. Note that with isolation parts you have to "read ahead" to find out what is being isolated if what one is doing with the parts is decorating a control or generating HTML.

Sure, but either the bidi isolation parts need to be included, or some way of obtaining their equivalent is required.